core/shared/src/main/scala-2/zio/internal/macros/LayerMacroUtils.scala

```diff
@@ -46,14 +46,60 @@ private[zio] trait LayerMacroUtils {
     val remainderTypes = getRequirements[R0]
     val targetTypes    = getRequirements[R]
     val debug          = typeOf[ZLayer.Debug.type].termSymbol
+    val trace          = c.freshName(TermName("trace"))
+    val compose        = c.freshName(TermName("compose"))
 
     val debugMap: PartialFunction[LayerExpr, ZLayer.Debug] = {
       case Expr(q"$prefix.tree") if prefix.symbol == debug    => ZLayer.Debug.Tree
       case Expr(q"$prefix.mermaid") if prefix.symbol == debug => ZLayer.Debug.Mermaid
     }
 
     def typeToNode(tpe: Type): Node[Type, LayerExpr] =
-      Node(Nil, List(tpe), c.Expr[ZLayer[_, E, _]](q"${reify(ZLayer)}.environment[$tpe]"))
+      Node(Nil, List(tpe), c.Expr[ZLayer[_, _, _]](q"${reify(ZLayer)}.environment[$tpe]($trace)"))
+
+    def buildFinalTree(tree: LayerTree[LayerExpr]): LayerExpr = {
+      val memoList: List[(LayerExpr, LayerExpr)] = tree.toList.map { node =>
+        val termName = c.freshName(TermName("layer"))
+        node -> c.Expr[ZLayer[_, _, _]](q"$termName")
+      }
+
+      val definitions = memoList.map { case (expr, memoizedNode) =>
+        q"val ${TermName(memoizedNode.tree.toString)} = $expr"
+      }
+
+      var needsComposeV = false
+      val memoMap       = memoList.toMap
+      val layerSym      = typeOf[ZLayer[_, _, _]].typeSymbol
+      val layerExpr = tree.fold[LayerExpr](
+        z = reify(ZLayer.unit),
+        value = memoMap,
+        composeH = (lhs, rhs) => c.Expr(q"$lhs ++ $rhs"),
+        composeV = (lhs, rhs) => {
+          needsComposeV = true
+          c.Expr(q"$compose($lhs, $rhs)")
+        }
+      )
+
+      val composeV = if (needsComposeV) {
+        val R  = c.freshName(TypeName("R"))
+        val E  = c.freshName(TypeName("E"))
+        val O1 = c.freshName(TypeName("O1"))
+        val O2 = c.freshName(TypeName("O2"))
+        List(q"""
+          def $compose[$R, $E, $O1, $O2](lhs: $layerSym[$R, $E, $O1], rhs: $layerSym[$O1, $E, $O2]) =
+            lhs.to(rhs)($trace)
+        """)
+      } else {
+        Nil
+      }
+
+      c.Expr(q"""
+        val $trace: ${typeOf[Trace]} = ${reify(Tracer)}.newTrace
```

Comment:
I believe this is going to bring back #8986, won't it?

Reply:
Now we use it also for `ZIO.environment` - I think this means always?

Reply:
I published locally your branch and it seems it does bring it back. Given the script below:

```scala
//> using scala "2.13.14"
//> using dep "dev.zio::zio::2.1.6+16-c49fcb17-SNAPSHOT"
//> using options -Xlint -Xfatal-warnings -Ywarn-macros:after -Wunused:locals

import zio._

object Foo {
  def foo = ZIO.unit.provide(ZLayer(ZIO.unit))
}
```

Running `scala-cli compile .` raises the following warning/error:

```
Compiling project (Scala 2.13.14, JVM (21))
[error] ./src/testy.scala:8:29
[error] local val trace$macro$1 in method foo is never used
[error]   def foo = ZIO.unit.provide(ZLayer(ZIO.unit))
[error]  
```

Reply:
PS: We probably need to set up a unit test for this, I'm just really not sure how to test for generated code :/

Reply:
Why would you provide a unit layer 😄

Reply:
Basically this is the issue `-Ywarn-macros:after` - too strict warnings

Reply:
@joroKr21 that was just for minimization. Providing unit layers is in fact a very common thing. e.g., this raises the same warning

```scala
//> using scala "2.13.14"
//> using dep "dev.zio::zio::2.1.6+16-c49fcb17-SNAPSHOT"
//> using options -Xlint -Xfatal-warnings -Ywarn-macros:after -Wunused:locals

import zio._

object Foo {
  def foo = ZIO.unit.provide(Runtime.enableCooperativeYielding)
}
```

Reply:
> Basically this is the issue -Ywarn-macros:after - too strict warnings

I know, but since users do use that compiler flag we shouldn't introduce a change that will break their code

Reply:
So we need to pull up that var flag - let me check if it can be done easily

Reply:
I added another mutable flag 🙈

Reply:
I just noticed the PR in my feed -- I must be following someone -- I haven't tried it yet, but the new normal is
```
 -Wmacros:default
```
which means only user code contributes definitions to check, and the macro expansion can only provide references that consume them.

Reply:
I didn't intend to become a reviewer, I was just sayin. 😅

Reply:
@som-snytt that's a good point, although unfortunately we still have to account for people using `-Ywarn-macros:after`. It's quite annoying for users to update the version of a library and get unused local value compiler errors totally unaware they're coming from ZIO's macros :/

> I didn't intend to become a reviewer, I was just sayin

And no worries, thanks for your input!